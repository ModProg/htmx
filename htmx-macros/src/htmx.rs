use htmx_script::{Script, ToJs};
use manyhow::{bail, Result};
use proc_macro2::{TokenStream, TokenTree};
use quote::ToTokens;
use rstml::atoms::OpenTag;
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, Node, NodeAttribute, NodeBlock,
    NodeElement, NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, ExprPath, LitStr};

use crate::*;

pub fn htmx(input: TokenStream) -> Result {
    // https://github.com/rust-lang/rust-analyzer/issues/15572
    // let htmx = match (
    //     proc_macro_crate::crate_name("htmx"),
    //     std::env::var("CARGO_CRATE_NAME").as_deref(),
    // ) { (Ok(FoundCrate::Itself), Ok("htmx")) => quote!(crate),
    //   (Ok(FoundCrate::Name(name)), _) => { let ident = Ident::new(&name,
    //   Span::call_site()); quote!(::#ident) } _ => quote!(::htmx),
    // };
    let htmx = htmx_crate();

    let mut input = input.into_iter().peekable();

    let htmx = match input.peek() {
        Some(TokenTree::Ident(ident)) if ident == "crate" => {
            input.next();
            quote!(crate)
        }
        _ => htmx,
    };

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .raw_text_elements(["script"].into()),
    )
    // TODO parse_recoverable
    .parse_simple(input.collect::<TokenStream>())?
    .into_iter()
    .map(expand_node)
    .collect::<Result<Vec<TokenStream>>>()?;
    let mut nodes = nodes.into_iter().peekable();

    Ok(if nodes.peek().is_some() {
        quote! {
        #use #htmx::{ToHtml, Html, IntoHtmlElements};
        {
            use #htmx::native::*;
            let mut $htmx = Html::new();
            #(
                ToHtml::write_to_html(&#nodes, &mut $htmx);
            )*
            $htmx
        }}
    } else {
        quote!(#htmx::Html::new())
    })
}

fn expand_node(node: Node) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let name = name_to_struct(name)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(_) => {
                        bail!(attribute, "dynamic attribute names not supported")
                    }
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!(.child(#script))
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!( .child({#[allow(unused_braces)] #block}))
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!(.child(#script))
                }
            } else {
                let children = children
                    .into_iter()
                    .map(expand_node)
                    .collect::<Result<Vec<_>>>()?;
                quote!(#(.child(#children))*)
            };
            let main = quote!(#name::builder() #(.#attributes)* #children .build());
            match close_tag.map(|tag| name_to_struct(tag.name)) {
                // If close_tag was specified, use it so coloring happens
                Some(Ok(close_tag)) if close_tag == name => quote!({let _ :#close_tag;#main}),
                _ => main,
            }
        }
        Node::Block(_) | Node::Text(_) => quote!( {#[allow(unused_braces)] #node}),
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => match c {},
    })
}

fn name_to_struct(name: NodeName) -> Result<ExprPath> {
    match name {
        NodeName::Path(path) => Ok(path),
        // This {...}
        NodeName::Punctuated(_) | NodeName::Block(_) => {
            bail!(name, "Only normal identifiers are allowd as node names")
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens) -> Result {
    match name {
        NodeName::Path(ExprPath { path, .. }) => Ok({
            let sident = path
                .segments
                .iter()
                .map(|i| i.ident.to_string().replace('_', "-"))
                .collect::<Vec<_>>()
                .join("-");
            if let Some(sident) = sident.strip_prefix("data-") {
                quote_spanned!(path.span()=> data(#sident, #value))
            } else if sident.starts_with("hx-") {
                quote_spanned!(path.span()=> data(#sident, #value))
            } else if let Some(ident) = path.get_ident() {
                quote!(#ident(#value))
            } else {
                bail!(path, "only `data::` or `hx::` are allowed as path prefix");
            }
        }),
        // This {...}
        NodeName::Punctuated(_) => {
            todo!("handle data-...")
        }
        NodeName::Block(_) => {
            bail!(
                name,
                "Only normal identifiers are allowd as attribute names"
            )
        }
    }
}
