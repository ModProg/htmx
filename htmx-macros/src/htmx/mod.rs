mod html;
mod special_components;

mod rusty;

pub use html::html;
pub use rusty::html as rtml;

// pub fn html(input: TokenStream) -> Result {
//     if input.is_empty() {
//         return Ok(quote!(::htmx::Html::new()));
//     }

//     let mut fork = input.clone().into_iter();

//     let first = fork.next();
//     let second = fork.next();

//     // TODO figure out actual differentiator
//     // probably would be, starts with `<` or starts with `{}` or `""` not
// followed     // by `,`

//     if matches!(input.peek(), Some(TokenTree::Punct(punct)) if
// punct.as_char() == '<') {         html::html(input.collect())
//     } else {
//         rusty::html(input.collect())
//     }
// }
