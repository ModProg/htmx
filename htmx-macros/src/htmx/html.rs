use htmx_script::{Script, ToJs};
use manyhow::{ensure, ErrorMessage, Result};
use proc_macro2::TokenStream;
use proc_macro_utils::TokenStream2Ext;
use quote::ToTokens;
use rstml::atoms::{CloseTag, OpenTag};
use rstml::node::{
    AttributeValueExpr, KeyedAttribute, KeyedAttributeValue, NodeAttribute, NodeBlock, NodeElement,
    NodeName,
};
use rstml::recoverable::Recoverable;
use syn::spanned::Spanned;
use syn::{parse2, Expr, ExprLit, ExprPath, Lit, LitStr, Stmt};

use super::special_components::{Node, Special};
use crate::*;

pub fn html(input: TokenStream) -> Result {
    let mut parser = input.clone().parser();
    let (input, expr) =
        if let ( Some(expr), Some(_)) = (parser.next_expression(), parser.next_tt_fat_arrow()) {
            (parser.into_token_stream(), quote!(&mut #expr))
        } else {
            (input, quote!(::htmx::Html::new()))
        };

    let nodes = rstml::Parser::new(
        rstml::ParserConfig::new()
            .recover_block(true)
            .element_close_use_default_wildcard_ident(false)
            .custom_node::<Special>()
            .raw_text_elements(["script"].into()),
    )
    // TODO parse_recoverable
    .parse_simple(input)?
    .into_iter()
    .map(expand_node)
    .collect::<Result>()?;

    Ok(quote! {
        {
            use ::htmx::native::*;
            let mut __html = #expr;
            #nodes
            __html
        }
    })
}

pub fn expand_node(node: Node) -> Result {
    Ok(match node {
        Node::Comment(_) => todo!("{}", line!()),
        Node::Doctype(_) => todo!("{}", line!()),
        Node::Fragment(_) => todo!("{}", line!()),
        Node::Element(NodeElement {
            open_tag: OpenTag {
                name, attributes, ..
            },
            children,
            close_tag,
            ..
        }) => {
            let script = name.to_string() == "script";
            let (name, custom) = name_to_struct(name)?;
            let attributes = attributes
                .into_iter()
                .map(|attribute| match attribute {
                    NodeAttribute::Block(attr) => Ok(quote!(custom_attr(#attr, true))),
                    NodeAttribute::Attribute(KeyedAttribute {
                        key,
                        possible_value,
                    }) => match possible_value {
                        KeyedAttributeValue::Binding(_) => todo!("{}", line!()),
                        KeyedAttributeValue::Value(AttributeValueExpr { value, .. }) => {
                            attribute_key_to_fn(key, value, custom)
                        }
                        KeyedAttributeValue::None => attribute_key_to_fn(key, true, custom),
                    },
                })
                .collect::<Result<Vec<_>>>()?;
            let children = if children.is_empty() {
                quote!()
            } else if script {
                // TODO scripts
                let Some(Node::RawText(script)) = children.first() else {
                    unreachable!("script always raw text")
                };
                let script = script.into_token_stream();
                if let Ok(script) = parse2::<LitStr>(script.clone()) {
                    quote!(__html.child_expr(#script);)
                } else if let Ok(block) =
                    parse2::<Recoverable<NodeBlock>>(script.clone()).map(Recoverable::inner)
                {
                    quote!(__html.child_expr({#[allow(unused_braces)] #block});)
                } else {
                    let script: Script = parse2(script)?;
                    let script = script.to_java_script();
                    quote!(__html.child(#script);)
                }
            } else {
                expand_nodes(children)?
            };
            let body = (!children.is_empty()).then(|| quote!(let mut __html = __html.body();));
            let main = quote!({let mut __html = #name #(__html.#attributes;)* #body; #children});

            match close_tag {
                Some(CloseTag {
                    name: name @ NodeName::Path(_),
                    ..
                }) if !name.is_wildcard() => {
                    // If close_tag was specified, use it so coloring happens
                    quote!({#name::unused(); #main})
                }
                _ => main,
            }
        }
        Node::Block(_) | Node::Text(_) => {
            quote!(::htmx::ToHtml::to_html(&{#[allow(unused_braces)] #node}, &mut __html);)
        }
        Node::RawText(_) => todo!("{}", line!()),
        Node::Custom(c) => c.expand_node()?,
    })
}

pub fn expand_nodes(nodes: Vec<Node>) -> Result {
    nodes.into_iter().map(expand_node).collect()
}

pub fn ensure_tag_name(name: String, span: impl ToTokens) -> Result<String, ErrorMessage> {
    ensure!(
        name.to_ascii_lowercase().chars()
            .all(|c| matches!(c, '-' | '.' | '0'..='9' | '_' | 'a'..='z' | '\u{B7}' | '\u{C0}'..='\u{D6}' | '\u{D8}'..='\u{F6}' | '\u{F8}'..='\u{37D}' | '\u{37F}'..='\u{1FFF}' | '\u{200C}'..='\u{200D}' | '\u{203F}'..='\u{2040}' | '\u{2070}'..='\u{218F}' | '\u{2C00}'..='\u{2FEF}' | '\u{3001}'..='\u{D7FF}' | '\u{F900}'..='\u{FDCF}' | '\u{FDF0}'..='\u{FFFD}' | '\u{10000}'..='\u{EFFFF}')),
        span,
        "invalid tag name `{name}`, https://html.spec.whatwg.org/multipage/custom-elements.html#prod-potentialcustomelementname"
        // TODO similar function but with css error: https://drafts.csswg.org/css-syntax-3/#non-ascii-ident-code-point
    );
    Ok(name)
}

fn name_to_struct(name: NodeName) -> Result<(TokenStream, bool)> {
    match name {
        NodeName::Path(path) => Ok((quote!(#path::new(&mut __html);), false)),
        name @ NodeName::Punctuated(_) => {
            let name = ensure_tag_name(name.to_string(), name)?;
            Ok((
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                true,
            ))
        }
        // This {...}
        NodeName::Block(name) => {
            if let [
                Stmt::Expr(
                    Expr::Lit(ExprLit {
                        lit: Lit::Str(name),
                        ..
                    }),
                    None,
                ),
            ] = &name.stmts[..]
            {
                let name = ensure_tag_name(name.value(), name)?;
                Ok((
                    quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);),
                    true,
                ))
            } else {
                Ok((
                    quote!(::htmx::CustomElement::new(&mut __html, #name);),
                    true,
                ))
            }
        }
    }
}

fn attribute_key_to_fn(name: NodeName, value: impl ToTokens, custom: bool) -> Result {
    Ok(match name {
        NodeName::Path(ExprPath { path, .. }) if !custom && path.get_ident().is_some() => {
            quote!(#path(#value))
        }
        NodeName::Path(ExprPath { path, .. })
            if path.segments.first().is_some_and(|hx| hx.ident == "hx") =>
        {
            {
                let sident = path
                    .segments
                    .iter()
                    .map(|i| i.ident.to_string().replace('_', "-"))
                    // hx::swap::oob
                    .collect::<Vec<_>>()
                    .join("-");
                quote_spanned!(path.span()=> custom_attr(#sident, #value))
            }
        }
        // This {...}
        name @ (NodeName::Punctuated(_) | NodeName::Path(_)) => {
            let sname = name.to_string();
            quote_spanned!(name.span()=>  custom_attr(#sname, #value))
        }
        name @ NodeName::Block(_) => quote_spanned!(name.span()=>  custom_attr(#name, #value)),
    })
}
