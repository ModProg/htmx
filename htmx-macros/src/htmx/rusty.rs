use std::mem;

use manyhow::{bail, ensure, Result};
use proc_macro2::{TokenStream, TokenTree};
use quote::{format_ident, ToTokens};
use syn::ext::IdentExt;
use syn::parse::discouraged::Speculative;
use syn::parse::{Parse, ParseStream, Parser, Peek};
use syn::punctuated::Punctuated;
use syn::token::{Brace, Bracket, Paren};
use syn::{bracketed, parenthesized, parse2, BinOp, Expr, LitStr, Pat, Path, Token};
use syn_derive::{Parse, ToTokens};

use super::html::ensure_tag_name;
use crate::*;

pub fn html(input: TokenStream) -> Result<proc_macro2::TokenStream, manyhow::Error> {
    let nodes = expand_nodes(Punctuated::<Node, Token![,]>::parse_terminated.parse2(input)?);

    Ok(quote! {
        #use ::htmx::{ToHtml, Html, IntoHtmlElements};
        {
            use ::htmx::native::*;
            let mut __html = Html::new();
            #(#nodes)*
            __html
        }
    })
}

fn expand_nodes(nodes: impl IntoIterator<Item = Node>) -> impl Iterator<Item = TokenStream> {
    nodes.into_iter().map(move |n| n.expand())
}

fn peek_alone(p: impl Peek, input: ParseStream) -> bool {
    input.peek(p)
        && (input.peek2(Token![,]) || {
            input.parse::<TokenTree>().unwrap();
            input.is_empty()
        })
}

#[derive(Debug, Parse)]
enum Node {
    #[parse(peek_func = |input| peek_alone(LitStr, input))]
    String(LitStr),
    #[parse(peek_func = |input| peek_alone(Brace, input))]
    Block(Block),
    #[parse(peek = Token![if])]
    If(If),
    #[parse(peek = Token![for])]
    For(For),
    #[parse(peek = Token![while])]
    While(While),
    // TODO controlflow
    Element(Element),
}

impl Node {
    fn expand(self) -> TokenStream {
        match self {
            Node::String(lit) => {
                quote!(::htmx::ToHtml::to_html(&#lit, &mut __html);)
            }
            Node::Block(block) => {
                quote!(::htmx::ToHtml::to_html(&#block, &mut __html);)
            },
            Node::Element(element) => element.expand(),
            Node::If(node) => node.expand(),
            Node::For(node) => node.expand(),
            Node::While(node) => node.expand(),
        }
    }
}

#[derive(Debug, Parse, ToTokens)]
struct Block {
    #[syn(braced)]
    brace: Brace,
    #[syn(in = brace)]
    content: TokenStream,
}

#[derive(Debug)]
struct If {
    if_token: Token![if],
    condition: Expr,
    #[allow(unused)]
    bracket: Option<Bracket>,
    then_branch: Punctuated<Node, Token![,]>,
    else_branch: ElseBranch,
}

impl If {
    fn expand(self) -> TokenStream {
        let If {
            if_token,
            condition,
            then_branch,
            else_branch,
            ..
        } = self;
        let body = expand_nodes(then_branch);
        let else_branch = else_branch.expand();
        quote! {
            #if_token #condition {
                #(#body)*
            } #else_branch
        }
    }
}

fn expr_before_bracket(input: ParseStream) -> syn::Result<Expr> {
    fn take_tt(input: ParseStream, output: &mut TokenStream) {
        output.extend(input.parse::<TokenTree>())
    }

    let expr = &mut TokenStream::new();
    'outer: while !input.is_empty() {
        if input.peek(Token![<]) {
            let mut turbo_fish: usize = 1;
            take_tt(input, expr);
            while turbo_fish > 0 {
                if input.peek(Token![>]) {
                    turbo_fish -= 1;
                    take_tt(input, expr);
                } else if input.peek(Token![<]) {
                    turbo_fish += 1;
                    take_tt(input, expr);
                } else if input.is_empty() {
                    return Err(input.error("expected `>` to close generic"));
                } else {
                    take_tt(input, expr);
                }
            }
        } else if let Ok(closure) = input.parse::<Token![|]>() {
            closure.to_tokens(expr);
            while !input.peek(Token![|]) {
                if input.is_empty() {
                    return Err(input.error("expected `|` to close closure arguments"));
                } else {
                    take_tt(input, expr)
                }
            }
        }
        while !input.is_empty() {
            if let Ok(path_sep) = input.parse::<Token![::]>() {
                path_sep.to_tokens(expr);
                continue 'outer;
            } else if let Ok(op) = input.parse::<BinOp>() {
                op.to_tokens(expr);
                continue 'outer;
            } else if let Ok(assign) = input.parse::<Token![=]>() {
                assign.to_tokens(expr);
                continue 'outer;
            } else if let Ok(let_) = input.parse::<Token![let]>() {
                let_.to_tokens(expr);
                Pat::parse_multi_with_leading_vert(input)?.to_tokens(expr);
            } else if input.peek(Bracket)
                && (input.peek2(Token![else]) || input.peek2(Token![,]) || peek2_eof(input))
            {
                return parse2(mem::take(expr));
            } else {
                take_tt(input, expr);
            }
        }
    }
    Err(input.error("expected `[` or `{`"))
}

fn peek2_eof(input: ParseStream) -> bool {
    let mut ts = input.fork().parse::<TokenStream>().unwrap().into_iter();
    ts.next();
    ts.next().is_none()
}

fn expr_for_cf(
    input: ParseStream,
) -> syn::Result<(Expr, Option<Bracket>, Punctuated<Node, Token![,]>)> {
    let fork = &input.fork();
    let condition = expr_before_bracket(fork);
    let error = match condition {
        Ok(condition) => {
            input.advance_to(fork);
            let content;
            return Ok((
                condition,
                Some(bracketed!(content in input)),
                Punctuated::parse_terminated(&content)?,
            ));
        }
        result => result.err(),
    };

    {
        let fork = &input.fork();
        if let Ok(condition) = Expr::parse_without_eager_brace(fork) {
            if let Ok(block) = fork.parse() {
                input.advance_to(fork);
                return Ok((condition, None, Punctuated::from_iter([Node::Block(block)])));
            }
        }
    }

    match error {
        Some(error) => bail!(error),
        _ => {
            input.advance_to(fork);
            bail!(input.error("expected `[` or `{`"));
        }
    }
}

impl Parse for If {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let if_token = input.parse()?;
        let (condition, bracket, then_branch) = expr_for_cf(input)?;
        Ok(Self {
            if_token,
            condition,
            bracket,
            then_branch,
            else_branch: input.parse()?,
        })
    }
}

#[derive(Debug)]
enum ElseBranch {
    None,
    Else {
        else_token: Token![else],
        #[allow(unused)]
        bracket: Option<Bracket>,
        body: Punctuated<Node, Token![,]>,
    },
    ElseIf {
        else_token: Token![else],
        body: Box<If>,
    },
}

impl ElseBranch {
    fn expand(self) -> TokenStream {
        match self {
            ElseBranch::None => quote!(),
            ElseBranch::Else {
                else_token, body, ..
            } => {
                let body = expand_nodes(body);
                quote!( #else_token {#(#body)*} )
            }
            ElseBranch::ElseIf { else_token, body } => {
                let body = body.expand();
                quote!(#else_token #body)
            }
        }
    }
}

impl Parse for ElseBranch {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(if let Ok(else_token) = input.parse() {
            if input.peek(Token![if]) {
                ElseBranch::ElseIf {
                    else_token,
                    body: Box::new(input.parse()?),
                }
            } else if input.peek(Bracket) {
                let body;
                ElseBranch::Else {
                    else_token,
                    bracket: Some(bracketed!(body in input)),
                    body: Punctuated::parse_terminated(&body)?,
                }
            } else if let Ok(block) = input.parse() {
                ElseBranch::Else {
                    else_token,
                    bracket: None,
                    body: Punctuated::from_iter([Node::Block(block)]),
                }
            } else {
                bail!(input.error("expected `[` or `{`"))
            }
        } else {
            ElseBranch::None
        })
    }
}

#[derive(Debug)]
struct For {
    for_token: Token![for],
    pat: syn::Pat,
    in_token: Token![in],
    expr: Expr,
    #[allow(unused)]
    bracket: Option<Bracket>,
    body: Punctuated<Node, Token![,]>,
}

impl For {
    fn expand(self) -> TokenStream {
        let Self {
            for_token,
            pat,
            in_token,
            expr,
            body,
            ..
        } = self;
        let body = expand_nodes(body);
        quote! {
            #for_token #pat #in_token #expr {
                #(#body)*
            }
        }
    }
}

impl Parse for For {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let for_token = input.parse()?;
        let pat = Pat::parse_single(input)?;
        let in_token = input.parse()?;

        let (expr, bracket, body) = expr_for_cf(input)?;
        Ok(Self {
            for_token,
            pat,
            in_token,
            expr,
            bracket,
            body,
        })
    }
}
#[derive(Debug)]
struct While {
    while_token: Token![while],
    expr: Expr,
    #[allow(unused)]
    bracket: Option<Bracket>,
    body: Punctuated<Node, Token![,]>,
}

impl While {
    fn expand(self) -> TokenStream {
        let Self {
            while_token,
            expr,
            body,
            ..
        } = self;
        let body = expand_nodes(body);
        quote! {
            #while_token #expr {
                #(#body)*
            }
        }
    }
}

impl Parse for While {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let while_token = input.parse()?;

        let (expr, bracket, body) = expr_for_cf(input)?;

        Ok(Self {
            while_token,
            expr,
            bracket,
            body,
        })
    }
}

#[derive(Debug)]
struct Element {
    path: ElementName,
    attrs: Option<Attrs>,
    #[allow(unused)]
    bracket: Option<Bracket>,
    children: Punctuated<Node, Token![,]>,
}
impl Element {
    fn expand(self) -> TokenStream {
        let mut attrs = self.attrs.unwrap_or_default();
        let name = match self.path {
            ElementName::String(name) => {
                quote!(::htmx::CustomElement::new_unchecked(&mut __html, #name);)
            }
            ElementName::Block(block) => quote!(::htmx::CustomElement::new(&mut __html, #block);),
            ElementName::Classes(classes) => {
                attrs.attrs.push(Attr::Classes(classes));
                quote!(div::new(&mut __html))
            }
            ElementName::Path(path) => quote!(#path::new(&mut __html);),
        };

        let mut children = expand_nodes(
            attrs
                .content
                .into_iter()
                .flat_map(|(_, c)| c)
                .chain(self.children),
        )
        .peekable();

        let attrs = attrs.attrs.into_iter().map(Attr::expand);

        let body = children.peek().is_some().then(|| quote!(__html.body(|mut __html| {#(#children)*})));

        quote!({
            let mut __html = #name
            #(__html #attrs;)* 
            #body;
        })
    }
}

impl Parse for Element {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let path = input.parse()?;
        let attrs = input.peek(Paren).then(|| input.parse()).transpose()?;

        let (bracket, children) = if input.peek(Bracket) {
            let nodes;
            (
                Some(bracketed!(nodes in input)),
                Punctuated::parse_terminated(&nodes)?,
            )
        } else {
            Default::default()
        };

        Ok(Self {
            path,
            attrs,
            bracket,
            children,
        })
    }
}

#[derive(Debug)]
enum ElementName {
    String(LitStr),
    Block(Block),
    Classes(Classes),
    Path(Path),
}

impl Parse for ElementName {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        if let Ok(lit) = input.parse::<LitStr>() {
            ensure_tag_name(lit.value(), &lit)?;
            Ok(Self::String(lit))
        } else if let Ok(block) = input.parse::<Block>() {
            parse2::<LitStr>(block.content.clone())
                .ok()
                .map_or(Ok(Self::Block(block)), |name| {
                    ensure_tag_name(name.value(), &name)
                        .map(|_| Self::String(name))
                        .map_err(Into::into)
                })
        } else if input.peek(Token![.]) {
            input.parse().map(Self::Classes)
        } else {
            input.parse().map(Self::Path)
        }
    }
}

#[derive(Debug, Parse)]
struct Classes {
    #[allow(unused)]
    leading: Token![.],
    #[parse(Punctuated::parse_separated_nonempty)]
    classes: Punctuated<Name, Token![.]>,
}

#[derive(Debug, Parse)]
enum Name {
    #[parse(peek = LitStr)]
    String(LitStr),
    #[parse(peek = Brace)]
    Block(Block),
    Ident(#[parse(Ident::parse_any)] Ident),
}

impl Name {
    fn lit_str(&self) -> Option<String> {
        match self {
            Name::String(lit) => Some(lit.value()),
            Name::Block(Block { content, .. }) => parse2::<LitStr>(content.clone())
                .ok()
                .as_ref()
                .map(LitStr::value),
            Name::Ident(_) => None,
        }
    }

    fn attribute(input: ParseStream) -> syn::Result<Self> {
        input.parse().and_then(|value: Self| {
            if let Some(value) = value.lit_str() {
                ensure!(
                    value.chars().any(|c| c.is_whitespace()
                        || c.is_control()
                        || matches!(c, '\0' | '"' | '\'' | '>' | '/' | '=')),
                    value,
                    "invalid key `{value}`, \
                    https://www.w3.org/TR/2011/WD-html5-20110525/syntax.html#attributes-0"
                );
            }
            Ok(value)
        })
    }
}

impl ToTokens for Name {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        match self {
            Name::String(lit) => lit.to_tokens(tokens),
            Name::Block(block) => block.to_tokens(tokens),
            Name::Ident(ident) => LitStr::new(&ident.to_string(), ident.span()).to_tokens(tokens),
        }
    }
}

#[derive(Debug, Default)]
struct Attrs {
    #[allow(unused)]
    paren: Paren,
    attrs: Punctuated<Attr, Token![,]>,
    content: Option<(Token![;], Punctuated<Node, Token![,]>)>,
}

impl Parse for Attrs {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let content;
        Ok(Self {
            paren: parenthesized!(content in input),
            attrs: {
                if !(content.peek(Token![;]) || content.is_empty()) {
                    let mut attrs = Punctuated::parse_separated_nonempty(&content)?;
                    if content.peek(Token![,]) {
                        attrs.push_punct(content.parse()?);
                    }
                    attrs
                } else {
                    Default::default()
                }
            },
            content: {
                if content.peek(Token![;]) {
                    Some((content.parse()?, Punctuated::parse_terminated(&content)?))
                } else {
                    None
                }
            },
        })
    }
}

#[derive(Debug)]
#[allow(unused)]
enum Attr {
    Id(Token![#], Name),
    Classes(Classes),
    // TODO Value(Expr),
    // TODO Flag(Name),
    // TODO StructShorthand(Name),
    KeyValue(Name, Token![:], Expr),
    Trailing(Token![..], Expr),
}

impl Parse for Attr {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(if input.peek(Token![#]) {
            Self::Id(input.parse()?, input.parse()?)
        } else if input.peek(Token![..]) {
            // This means, that attribute names need to use parentheses to be lhs open
            // ranges
            Self::Trailing(input.parse()?, input.parse()?)
        } else if input.peek(Token![.]) {
            Self::Classes(input.parse()?)
        } else {
            // let expr = input.parse()?;
            // if input.peek(Token![:]) {
            Self::KeyValue(input.call(Name::attribute)?, input.parse()?, input.parse()?)
            // } else {
            // Self::Value(expr)
            // }
        })
    }
}

fn is_keyword(ident: &Ident) -> bool {
    match ident.to_string().as_str() {
        // Based on https://doc.rust-lang.org/1.65.0/reference/keywords.html
        "abstract" | "as" | "async" | "await" | "become" | "box" | "break" | "const"
        | "continue" | "crate" | "do" | "dyn" | "else" | "enum" | "extern" | "false" | "final"
        | "fn" | "for" | "if" | "impl" | "in" | "let" | "loop" | "macro" | "match" | "mod"
        | "move" | "mut" | "override" | "priv" | "pub" | "ref" | "return" | "Self" | "self"
        | "static" | "struct" | "super" | "trait" | "true" | "try" | "type" | "typeof"
        | "unsafe" | "unsized" | "use" | "virtual" | "where" | "while" | "yield" => true,
        _ => false,
    }
}

impl Attr {
    fn expand(self) -> TokenStream {
        match self {
            Attr::Id(_, id) => quote!(.id(#id)),
            Attr::Classes(classes) => {
                let classes = classes.classes.into_iter();
                quote!(#(.class(#classes))*)
            }
            Attr::KeyValue(name, _, value) => match name {
                Name::Ident(ref name) if is_keyword(name) => {
                    let name = format_ident!("{name}_");
                    quote!(.#name(#value))
                }
                Name::Ident(name) => quote!(.#name(#value)),
                name => {
                    if name.lit_str().is_some() {
                        quote!(.custom_attr_unchecked(#name, #value))
                    } else {
                        quote!(.custom_attr(#name, #value))
                    }
                }
            },
            Attr::Trailing(..) => todo!(),
        }
    }
}

#[cfg(test)]
mod test {
    use forr::forr;
    use proc_macro_utils::assert_tokens;

    use super::*;

    #[test]
    fn expr_before_bracket() {
        fn helper(input: ParseStream) -> syn::Result<(Expr, TokenStream)> {
            Ok((super::expr_before_bracket(input)?, input.parse()?))
        }

        forr! { $expr:inner in [(1 + 1), (<[u8]>::hello::<[u8], Hello<[u8],> >)] $*
        forr! { $after:inner in [([],), ([] else), ([])] $*
            let tokens = quote!($expr $after);
            let (expr, rest) = helper.parse2(tokens).unwrap();
            assert_tokens!(expr.into_token_stream(), {$expr});
            assert_tokens!(rest.into_token_stream(), {$after});
        }}
    }
}
