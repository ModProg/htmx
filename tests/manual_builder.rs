use htmx::{Children, Html, ToHtml};

#[derive(htmx::__private::typed_builder::TypedBuilder)]
#[builder(build_method(into = htmx::Html))]
struct Component {
    #[builder(via_mutators, mutators(
        fn child(&mut self, child: impl htmx::ToHtml) {
           self.children.push(child);
        }
    ))]
    children: Children,
}

impl From<Component> for Html {
    fn from(Component { children }: Component) -> Self {
        let mut html = Html::new();
        htmx::native::div::builder()
            .child(children)
            .build()
            .write_to_html(&mut html);
        html
    }
}

#[test]
fn test() {
    insta::assert_snapshot!(
        Component::builder()
            .child("hello")
            .child("world")
            .build()
            .to_string()
    )
}
