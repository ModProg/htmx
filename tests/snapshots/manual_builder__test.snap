---
source: tests/manual_builder.rs
expression: "Component::builder().child(\"hello\").child(\"world\").build().to_string()"
---
<!DOCTYPE html><div> hello  world </div>
